com-api/com-api-concept/src/lib.rs

```diff
@@ -160,48 +239,83 @@ where
     ///
     /// This corresponds to `MaybeUninit::write`.
     fn write(self, value: T) -> Self::SampleMut;
+
+    /// Get a mutable pointer to the internal maybe uninitialized `T`.
+    ///
+    /// The caller has to make sure to initialize the data in the buffer.
+    /// Reading from the received pointer before initialization is undefined behavior.
+    fn as_mut_ptr(&mut self) -> *mut T;
 }
 
-pub trait Interface {}
+pub trait Interface {
+    type Consumer<R: Runtime + ?Sized>: Consumer<R>;
+    type Producer<R: Runtime + ?Sized>: Producer<R>;
+}
 
-pub trait OfferedProducer {
+#[must_use = "if a service is offered it will be unoffered and dropped immediately, causing unexpected behavior in the system"]
+pub trait OfferedProducer<R: Runtime + ?Sized> {
     type Interface: Interface;
-    type Producer: Producer<Interface = Self::Interface>;
+    type Producer: Producer<R, Interface = Self::Interface>;
 
     fn unoffer(self) -> Self::Producer;
 }
 
-pub trait Producer {
+pub trait Producer<R: Runtime + ?Sized> {
     type Interface: Interface;
-    type OfferedProducer: OfferedProducer<Interface = Self::Interface>;
+    type OfferedProducer: OfferedProducer<R, Interface = Self::Interface>;
 
     fn offer(self) -> Result<Self::OfferedProducer>;
 }
 
-pub trait Consumer {}
+pub trait Publisher<T>
+where
+    T: Reloc + Send + Debug,
+{
+    type SampleMaybeUninit<'a>: SampleMaybeUninit<T> + 'a
+    where
+        Self: 'a;
+
```

Comment:
We will need new here
```suggestion
fn new(identifier: &str, instance_info: R::ProviderInfo) -> Self;
```

Reply:
The chain here should be `Runtime::producer_builder().build().offer()`.

Reply:
Yes, it is needed to be able to crate Publisher with identifier and instance info passed.
In current implementation there is no way to provide service name for Publisher.